src/codeblockediting.js

```diff
@@ -0,0 +1,409 @@
+				return false;
+			}
+		} );
+
+		// Disallow all attributes in `codeBlock`.
+		schema.addAttributeCheck( ( context, attributeName ) => {
+			if ( context.endsWith( 'codeBlock' ) || context.endsWith( 'codeBlock $text' ) ) {
+				return attributeName === 'language';
+			}
+		} );
+
+		// Conversion.
+		editor.editing.downcastDispatcher.on( 'insert:codeBlock', modelToViewCodeBlockInsertion( model, languageLabels ) );
+		editor.data.downcastDispatcher.on( 'insert:codeBlock', modelToViewCodeBlockInsertion( model ) );
+		editor.data.downcastDispatcher.on( 'insert:softBreak', modelToDataViewSoftBreakInsertion( model ), { priority: 'high' } );
+		editor.data.upcastDispatcher.on( 'element:pre', dataViewToModelCodeBlockInsertion( editor.data, languageClasses ) );
+
+		// Intercept the clipboard input (paste) when the selection is anchored in the code block and force the clipboard
+		// data to be pasted as a single plain text. Otherwise, the code lines will split the code block and
+		// "spill out" as separate paragraphs.
+		this.listenTo( editor.editing.view.document, 'clipboardInput', ( evt, data ) => {
+			const modelSelection = model.document.selection;
+
+			if ( !modelSelection.anchor.parent.is( 'codeBlock' ) ) {
+				return;
+			}
+
+			const text = data.dataTransfer.getData( 'text/plain' );
+
+			model.change( writer => {
+				model.insertContent( rawSnippetTextToModelDocumentFragment( writer, text ), modelSelection );
+				evt.stop();
+			} );
+		} );
+
+		// Make sure multi–line selection is always wrapped in a code block when `getSelectedContent()`
+		// is used (e.g. clipboard copy). Otherwise, only the raw text will be copied to the clipboard and,
+		// upon next paste, this bare text will not be inserted as a code block, which is not the best UX.
+		// Similarly, when the selection in a single line, the selected content should be an inline code
+		// so it can be pasted later on and retain it's preformatted nature.
+		this.listenTo( model, 'getSelectedContent', ( evt, [ selection ] ) => {
+			const anchor = selection.anchor;
+
+			if ( selection.isCollapsed || !anchor.parent.is( 'codeBlock' ) || !anchor.hasSameParentAs( selection.focus ) ) {
+				return;
+			}
+
+			model.change( writer => {
+				const docFragment = evt.return;
+
+				// fo[o<softBreak></softBreak>b]ar  ->   <codeBlock language="...">[o<softBreak></softBreak>b]<codeBlock>
+				if ( docFragment.childCount > 1 || selection.containsEntireContent( anchor.parent ) ) {
+					const codeBlock = writer.createElement( 'codeBlock', anchor.parent.getAttributes() );
+					writer.append( docFragment, codeBlock );
+
+					const newDocumentFragment = writer.createDocumentFragment();
+					writer.append( codeBlock, newDocumentFragment );
+
+					evt.return = newDocumentFragment;
+				}
+
+				// "f[oo]"                          ->   <$text code="true">oo</text>
+				else {
+					writer.setAttribute( 'code', true, docFragment.getChild( 0 ) );
```

Comment:
self.r-: We must check first if the code attribute is allowed (check if the Code plugin is loaded).